src/storage/src/types/errors.rs

```diff
@@ -50,22 +50,200 @@ impl Codec for DecodeError {
+        use proto_envelope_error::Kind;
+        match proto.kind {
+            Some(Kind::Debezium(text)) => Ok(Self::Debezium(text)),
+            Some(Kind::Upsert(proto)) => {
+                let rust = RustType::from_proto(*proto)?;
+                Ok(Self::Upsert(rust))
+            }
+            Some(Kind::Flat(text)) => Ok(Self::Flat(text)),
+            None => Err(TryFromProtoError::missing_field("ProtoEnvelopeError::kind")),
+        }
+    }
+}
+
+impl Display for EnvelopeError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            DecodeError::Text(e) => write!(f, "Text: {}", e),
+            EnvelopeError::Debezium(err) => write!(f, "Debezium: {err}"),
+            EnvelopeError::Upsert(err) => write!(f, "Upsert: {err}"),
+            EnvelopeError::Flat(err) => write!(f, "Flat: {err}"),
+        }
+    }
+}
+
+/// An error from a value in an upsert source. The corresponding key is included, allowing
+/// us to reconstruct their entry in the upsert map upon restart.
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub struct UpsertValueError {
+    /// The underlying error.
+    pub inner: Box<DataflowError>,
+    /// The (good) key associated with the errored value.
+    pub for_key: Row,
+}
+
+impl RustType<ProtoUpsertValueError> for UpsertValueError {
+    fn into_proto(&self) -> ProtoUpsertValueError {
+        ProtoUpsertValueError {
+            inner: Some(self.inner.into_proto()),
+            for_key: Some(self.for_key.into_proto()),
+        }
+    }
+
+    fn from_proto(proto: ProtoUpsertValueError) -> Result<Self, TryFromProtoError> {
+        let inner = match proto.inner {
+            Some(inner) => RustType::from_proto(inner)?,
+            None => {
+                return Err(TryFromProtoError::missing_field(
+                    "ProtoUpsertValueError::inner",
+                ))
+            }
+        };
+        let for_key = match proto.for_key {
+            Some(key) => RustType::from_proto(key)?,
+            None => {
+                return Err(TryFromProtoError::missing_field(
+                    "ProtoUpsertValueError::for_key",
+                ))
+            }
+        };
+        Ok(Self { inner, for_key })
+    }
+}
+
+impl Display for UpsertValueError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let UpsertValueError { inner, for_key } = self;
+        write!(f, "{inner}, decoded key: {for_key:?}")
+    }
+}
+
+/// An error that can be retracted by a future message using upsert logic.
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub enum UpsertError {
+    /// Wrapper around a key decoding error.
+    /// We use this instead of emitting the underlying `DataflowError::DecodeError` because with only
+    /// the underlying error, we can't distinguish between an error with the key and an error
+    /// with the value.
```

Comment:
maybe add some more color on why we need to distinguish between these? its still unclear to me why

Reply:
Added a comment.